File: kmeans.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <limits>

using namespace std;

class Point {
public:
    vector<double> coords;

    Point(const vector<double>& coords = {}) : coords(coords) {}
};

class Cluster {
public:
    Point center;
    vector<Point> points;
    int dimensions;

    //constructor
    Cluster(const Point& center) : center(center), dimensions(center.coords.size()) {}

    void addPoint(const Point& p) {
        points.push_back(p);
    }

    void clear() {
        points.clear();
    }

    //updates center of cluster
    void update() {
        vector<double> sum(dimensions, 0);
        for (const auto& p : points) {
            for (int i = 0; i < dimensions; ++i) {
                sum[i] += p.coords[i];
            }
        }
        for (int i = 0; i < dimensions; ++i) {
            center.coords[i] = sum[i] / points.size();
        }
    }

    bool isEmpty() const {
        return points.empty();
    }
};

double distance(const Point& p1, const Point& p2) {
    double sum = 0;
    for (int i = 0; i < p1.coords.size(); ++i) {
        double d = p1.coords[i] - p2.coords[i];
        sum += d * d;
    }
    return sqrt(sum);
}

int findNearestCluster(const vector<Cluster>& clusters, const Point& p) {
    int index = 0;
    double minDistance = numeric_limits<double>::max();
    for (int i = 0; i < clusters.size(); ++i) {
        double d = distance(clusters[i].center, p);
        if (d < minDistance) {
            index = i;
            minDistance = d;
        }
    }
    return index;
}

vector<Cluster> kmeans(const vector<Point>& points, int k, int maxIterations) {
    vector<Cluster> clusters(k, Cluster(Point(vector<double>(points[0].coords.size(), 0))));

    // Initialize clusters with random points
    for (int i = 0; i < k; ++i) {
        int randomIndex = rand() % points.size();
        clusters[i].center = points[randomIndex];
    }

    // Assign points to initial clusters
    for (const auto& p : points) {
        int nearestCluster = findNearestCluster(clusters, p);
        clusters[nearestCluster].addPoint(p);
    }

    // Update clusters until convergence or maxIterations is reached
    int iteration = 0;
    while (iteration < maxIterations) {
        bool converged = true;
        for (auto& c : clusters) {
            if (!c.isEmpty()) {
                Point oldCenter = c.center;
                c.update();
                if (distance(oldCenter, c.center) > 1e-9) {
                    converged = false;
                }
            }
        }
        if (converged) {
            break;
        }
	//clear out clusters and reassin points
        for (auto& c : clusters) {
            c.clear();
        }
        for (const auto& p : points) {
            int nearestCluster = findNearestCluster(clusters, p);
            clusters[nearestCluster].addPoint(p);
        }
        ++iteration;
    }

    return clusters;
}

int main() {
    vector<Point> points = {Point({1, 2, -3}), Point({3, 4, 0}), Point({5, 6, 1}), Point({7, 3, -1})};
    int k = 2;
    int maxIterations = 100;
	
    // Run k-means clustering
	vector<Cluster> clusters = kmeans(points, k, maxIterations);
	
	// Print cluster centers and points
	for (int i = 0; i < clusters.size(); ++i) {
		cout << "Center of cluster " << i << ": (";
		for (int j = 0; j < clusters[i].center.coords.size(); ++j) {
			if (j > 0) {
				cout << ", ";
			}
			cout << clusters[i].center.coords[j];
		}
		cout << ")" << endl;
		cout << "Points in cluster " << i << ":" << endl;
		for (const auto& p : clusters[i].points) {
			cout << "(";
			for (int j = 0; j < p.coords.size(); ++j) {
				if (j > 0) {
					cout << ", ";
				}
				cout << p.coords[j];
			}
			cout << ")" << endl;
		}
	}

	return 0;

}

File: kmeans.ml
open Random

type point = float array

type cluster = { mutable center : point; mutable points : point list }

let distance p1 p2 =
  let sum_sq_diff = ref 0.0 in
  Array.iter2 (fun x y -> sum_sq_diff := !sum_sq_diff +. (x -. y) ** 2.0) p1 p2;
  sqrt !sum_sq_diff

let find_nearest_cluster point clusters =
  let nearest_cluster = ref None in
  let min_distance = ref infinity in
  List.iter (fun cluster ->
    let d = distance point cluster.center in
    if d < !min_distance then (
      nearest_cluster := Some cluster;
      min_distance := d
    )
  ) clusters;
  match !nearest_cluster with
  | Some cluster -> cluster
  | None -> failwith "No clusters"

let rec update_clusters points clusters iterations max_iterations =
  (*clear clusters and reassing points*)
  List.iter (fun cluster -> cluster.points <- []) clusters;
  List.iter (fun point ->
    let nearest_cluster = find_nearest_cluster point clusters in
    nearest_cluster.points <- point :: nearest_cluster.points
  ) points;
  let converged = ref true in
  List.iter (fun cluster ->
    let old_center = cluster.center in
    match cluster.points with
    | [] -> ()
    | _ ->
      (*calculates new center for clusters*)
      let dim = Array.length old_center in
      let new_center = Array.make dim 0.0 in
      List.iter (fun point ->
        Array.iteri (fun i coord -> new_center.(i) <- new_center.(i) +. coord) point
      ) cluster.points;
      Array.iteri (fun i _ ->
        new_center.(i) <- new_center.(i) /. float_of_int (List.length cluster.points)
      ) new_center;
      cluster.center <- new_center;
      if not (old_center = new_center) then converged := false
  ) clusters;
  (*loop if not converged*)
  if not !converged && iterations + 1 < max_iterations then update_clusters points clusters (iterations + 1) max_iterations

let kmeans points k max_iterations =
  let clusters = Array.make k { center = [||]; points = [] } in
  (*init clusters with random points*)
  for i = 0 to k - 1 do
    clusters.(i) <- { center = Array.copy points.(i); points = [] }
  done;
  for i = k to Array.length points - 1 do
    let random_cluster = clusters.(Random.int k) in
    random_cluster.points <- points.(i) :: random_cluster.points
  done;
  update_clusters (Array.to_list points) (Array.to_list clusters) 0 max_iterations;
  Array.to_list clusters

let () =
  let points = [|
    [| 1.0; 2.0; 3.0 |];
    [| 4.0; 5.0; 6.0 |];
    [| 7.0; 8.0; 9.0 |];
    [| 2.0; 3.0; 4.0 |];
    [| 5.0; 6.0; 7.0 |];
    [| 8.0; 9.0; 10.0 |];
  |] in
  let k = 2 in
  let max_iterations = 10 in
  let clusters = kmeans points k max_iterations in
  List.iteri (fun i cluster ->
    Printf.printf "Center of cluster %d: (%s)\n" i (Array.fold_left (fun acc x -> acc ^ (string_of_float x) ^ ", ") "" cluster.center);
	Printf.printf "Points in cluster %d: \n" i;
    List.iter (fun point ->
		Printf.printf "(%s)\n" (Array.fold_left (fun acc x -> acc ^ (string_of_float x) ^ ", ") "" point)) cluster.points) clusters;
